sphinx/ext/autodoc.py

```diff
@@ -107,49 +106,96 @@ def __getattr__(self, name):
             return None
 
 
-class _MockModule(object):
+class _MockObject(object):
     """Used by autodoc_mock_imports."""
-    __file__ = '/dev/null'
-    __path__ = '/dev/null'
 
     def __init__(self, *args, **kwargs):
         # type: (Any, Any) -> None
-        self.__all__ = []  # type: List[str]
+        pass
+
+    def __len__(self):
+        # type: () -> int
+        return 0
+
+    def __contains__(self, key):
+        # type: (str) -> bool
+        return False
 
-    def __call__(self, *args, **kwargs):
-        # type: (Any, Any) -> _MockModule
+    def __iter__(self):
+        # type: () -> None
+        pass
+
+    def __getitem__(self, key):
+        # type: (str) -> _MockObject
+        return self
+
+    def __getattr__(self, key):
+        # type: (str) -> _MockObject
+        return self
+
+    def __call__(self, *args, **kw):
+        # type: (Any, Any) -> Any
         if args and type(args[0]) in [FunctionType, MethodType]:
             # Appears to be a decorator, pass through unchanged
             return args[0]
-        return _MockModule()
+        return self
 
-    def _append_submodule(self, submod):
-        # type: (str) -> None
-        self.__all__.append(submod)
 
-    @classmethod
-    def __getattr__(cls, name):
-        # type: (unicode) -> Any
-        if name[0] == name[0].upper():
-            # Not very good, we assume Uppercase names are classes...
-            mocktype = type(name, (), {})  # type: ignore
-            mocktype.__module__ = __name__
-            return mocktype
-        else:
-            return _MockModule()
+class _MockModule(ModuleType):
+    """Used by autodoc_mock_imports."""
+    __file__ = '/dev/null'
+    __path__ = '/dev/null'
 
+    def __init__(self, name):
+        # type: (str) -> None
+        self.__name__ = name
+        self.__all__ = []  # type: List[str]
+
+    def __getattr__(self, name):
+        # type: (str) -> _MockObject
+        o = _MockObject()
+        o.__module__ = self.__name__
+        return o
+
+
+class _MockImportHook(object):
+
+    def __init__(self, names):
+        # type: (List[str]) -> None
+        self.packages = set()  # type: Set[str]
+        for n in names:
+            # Convert the list of module names:
+            #     ['a.b.c', 'd.e']
+            # to a set of base packages:
+            #     set(['a', 'd'])
+            self.packages.add(n.split('.')[0])
```

Comment:
Yeah I understand your concern. The problem is that in order to mock `a.b`, you need to also mock `a`. This is explained in [PEP 302](https://www.python.org/dev/peps/pep-0302/):

> The Importer Protocol operates at this level of individual imports. By the time an importer gets a request for "spam.ham", module "spam" has already been imported.

If you want, I can do a more complex algorithm to make sure that if `autodoc_mock_imports = ['a.b']`, only `a`, `a.b` and sub packages (`a.b.*`) are mocked and not `a.c`.

But since I remove the mocked modules from `sys.modules` after running the auto documenter, this should not have any consequences in Sphinx overall.